Erstmal Daten einlesen: Die Datei all.txt enthält den kompletten Text meines Datenbank-Buchs https://leanpub.com/realtionaledatenbanken/

In [1]:
textFile = sc.textFile("all.txt")
textFile.count()


2427

Wir brauchen eine Hilfsfunktion, die die Zeilen in einzelne Wörter zerlegt.
Hier könnte man dann auch noch ein bisschen "sauber machen", also etwa Formatierungszeichen
entfernen, Schreibweisen korrigieren, Umlaute entfernen usw.

In [3]:
def parse_line(line):
    return line.split(" ")

Jetzt geht's los: Zunächst mal erzeugen wir ein RDD, das alle Wörter des Texts enthält. Die Datei enthölt 2427 Zeilen und 32674 Wörter

In [7]:
words=textFile.flatMap(parse_line)
print textFile.count()
print words.count()

2427
32674


Jetzt machen wir uns ein pair RDD, das jedes Wort und die bisherige Anzahl (1) enthält

In [8]:
initial_count = words.map(lambda word: (word, 1))
print initial_count.take(5)

[(u'#', 1), (u'Einleitung', 1), (u'Datenbanken', 1), (u'sind', 1), (u'aus', 1)]


Das war die Map-Phase. Jetzt fassen wir die gleichen Wörter zusammen und zählen:

In [10]:
word_count = initial_count.reduceByKey(lambda a, b: a+b)
print word_count.take(10)

[(u'', 5264), (u'Punkten', 1), (u'Basiert', 1), (u'Sprache', 1), (u'(select)', 1), (u'gleichzeitig,', 1), (u'Schl\xfcsselattribut', 1), (u'validates_uniqueness_of', 1), (u'Transaktion', 28), (u'gerade', 1)]


Oder alles auf einmal:

In [16]:
word_count = (textFile
    .flatMap(parse_line)
    .map(lambda word: (word, 1))
    .reduceByKey(lambda a, b: a+b))


Wir wollen natürlich die häufigsten Wörter sehen ... wir sortieren nach der negativen Anzahl, um die Wörter in absteigender Anzahl zu erhalten

In [17]:
print word_count.takeOrdered(50, key = lambda x: -x[1])

[(u'', 5264), (u'die', 768), (u'der', 651), (u'in', 339), (u'und', 335), (u'ist', 280), (u'von', 275), (u'auf', 252), (u'eine', 249), (u'|', 232), (u'werden', 230), (u'wird', 212), (u'ein', 201), (u'zu', 192), (u'den', 190), (u'das', 174), (u'nicht', 171), (u'einer', 168), (u'mit', 155), (u'werden.', 152), (u'f\xfcr', 137), (u'bei', 136), (u'Daten', 133), (u'werden,', 127), (u'Die', 127), (u'etwa', 127), (u'=', 127), (u'sind', 125), (u'dass', 124), (u'oder', 123), (u'im', 119), (u'sich', 114), (u'es', 113), (u'als', 113), (u'nur', 105), (u'kann', 100), (u'Datenbank', 99), (u'auch', 97), (u'des', 94), (u'einem', 92), (u'einen', 90), (u'diese', 89), (u'Datenbanken', 89), (u'k\xf6nnen', 85), (u'wie', 84), (u'aber', 84), (u'zur', 83), (u'dann', 79), (u'aus', 78), (u'wenn', 73)]


In [ ]:
Mal schauen, wieviele Woerter nur einmal vorkommen:

In [15]:
singletons = word_count.filter(lambda (word,count): count == 1)
print singletons.count()

4709
